hadoop-hdds/docs/content/design/storage-class.md

```diff
@@ -19,10 +19,331 @@ author: Marton Ele
+
+When keys are created, blocks are assigned from the appropriate containers. **There is no way to change storage-class without moving data* (distcp).
+
+bucket.storage-class := STANDARD --> REDUCED
+
+container=REDUCED
+
+key1 (storage-class=REDUCED)
+key2 (straoge-class=REDUCED)
+
+REDUCED: (RATIS/ONE -- CLOSED/ONE) --> (RATIS/TWO -- CLOSED/TWO)
+
+## Multi-stage storage-class
+
+There is a very specific use case the "temperature" of data. Key (and containers) can become COLD or HOT over the the time. As we need transitions between the different **state** of the containers, the COLD data should be a new state of a container not a storage class.
+
+Therefore the STANDARD storage class can be modified to support three states: OPEN, WARM(=CLOSED), COLD. Transition between WARM and CLOSED can be managed by SCM (if container is not used, SCM can request to enable erasure coding). Or manually (one specific container can be forced to Erasure Coding / COLD state. Only useful for administrators)
+
+![Storage class transitions](storage-class-transition.png)
+
+## Storage class on the user interface
+
+Today, we need to define the replication factor (like `THREE`) and replication type (like `RATIS`) during the creation of the key. It's not possible to adjust possible replication rule for the key when it become part of a closed containers.
+
+With storage-class support the users don't need to udnerstand the details of the replication. Users can choose from the well-known or custom-defined storage-classes.
+
+# Implementation changes
+
+**Phase 1**: while storage class abstraction can provide a very generic framework for replication, in the first implementation It's proposed to keep it as simple as possible:
+
+ * Don't introduce any new custom container state
+ * Don't introduce any new transitions
+ * Don't introduce any new custom storage classes
+
+But we can start to use the abstraction for the existing implementation:
+
+ * We will implement two, hard-coded storage-classes (STANDARD, REDUCED) which can configure the behavior of existing replication schemes
+ * Ozone Client will be changed to support `--storage-class` instead of `--type` and `--factor`
+ * `ofs://` and `o3fs://` doesn't require any modification
+ * Storage class will be used as a custom String in RPC protocol (in storage-sensitive places it can be stored as an integer with an additional mapping)
+ * All the replication logic (PipelineManager/ReplicationManager) will work exactly as before. Storage-class will be resolved to the required replication config. Pipelines will have the same type as before (eg. Ratis/THREE)
+ * SCM allocateContainer logic should be changed. Right now it creates a predefined number of Open containers for each `owner` (high level application). It can be extended easily to create open containers for each storage-classes
```

Comment:
Please add more implementation details - how do you propose to extend the SCM? What parts of the SCM will change?

Reply:
Thanks for the suggestion. I extended this part with different pagraph for OM, SCM, Client... changes (instead of using only one lines).